Sources/TuistCache/Cache/FileClient.swift

```diff
@@ -0,0 +1,147 @@
+
+    // MARK: - Init
+
+    public init(session: URLSession = URLSession.shared) {
+        self.session = session
+    }
+
+    // MARK: - Public
+
+    public func download(url: URL) -> Single<AbsolutePath> {
+        dispatchDownload(request: URLRequest(url: url)).map { AbsolutePath($0.path) }
+    }
+
+    public func upload(file: AbsolutePath, hash _: String, to url: URL) -> Single<Bool> {
+        Single<Bool>.create { observer -> Disposable in
+            do {
+                let fileSize = try FileHandler.shared.fileSize(path: file)
+                let fileData = try Data(contentsOf: file.url)
+
+                let request = self.uploadRequest(url: url, fileSize: fileSize, data: fileData)
+                let uploadTask = self.session.dataTask(with: request) { _, response, error in
+                    if let error = error {
+                        observer(.error(FileClientError.urlSessionError(error, file)))
+                    } else if let response = response as? HTTPURLResponse {
+                        if self.successStatusCodeRange.contains(response.statusCode) {
+                            observer(.success(true))
+                        } else {
+                            observer(.error(FileClientError.serverSideError(response, file)))
+                        }
+                    } else {
+                        observer(.error(FileClientError.invalidResponse(request, file)))
+                    }
+                }
+                uploadTask.resume()
+                return Disposables.create { uploadTask.cancel() }
+            } catch {
+                observer(.error(error))
+            }
+            return Disposables.create {}
+        }
+    }
+
+    // MARK: - Private
+
+    private func uploadRequest(url: URL, fileSize: UInt64, data: Data) -> URLRequest {
+        var request = URLRequest(url: url)
+        request.httpMethod = "PUT"
+        request.setValue("application/zip", forHTTPHeaderField: "Content-Type")
+        request.setValue(String(fileSize), forHTTPHeaderField: "Content-Length")
+        request.setValue("zip", forHTTPHeaderField: "Content-Encoding")
+        request.httpBody = data
+        return request
+    }
+
+    private func dispatchDownload(request: URLRequest) -> Single<URL> {
+        Single.create { observer in
+            let task = self.session.downloadTask(with: request) { localURL, response, networkError in
+                if let networkError = networkError {
+                    observer(.error(FileClientError.urlSessionError(networkError, nil)))
+                } else if let response = response as? HTTPURLResponse {
+                    guard let localURL = localURL else {
+                        observer(.error(FileClientError.noLocalURL(response)))
+                        return
+                    }
+
+                    if self.successStatusCodeRange.contains(response.statusCode) {
+                        observer(.success(localURL))
```

Comment:
Keep in mind thas URLSession deletes the file after the completion block exits. This might be a problem in the future. Someone trying to copy that file somewhere else, and the file not existing because URLSession has decided to delete it. I'd recommend moving it to a temporary directory in this closure and then passing the new path instead of giving the consumer the option to delete it whenever they are done with the file.

Reply:
Addressed.

Reply:
Tio, I'd keep it simple like it is today and modify it whenever we need a more custom solution. if that's okay I'll merge this PR, I think I addressed all the other comments ;)